File: examples/python/servers/advanced/bazaar.py
"""Bazaar discovery extension example."""

import json
import logging
import os

from dotenv import load_dotenv
from fastapi import FastAPI
from pydantic import BaseModel

from x402.extensions.bazaar import (
    OutputConfig,
    bazaar_resource_server_extension,
    declare_discovery_extension,
)
from x402.http import FacilitatorConfig, HTTPFacilitatorClient, PaymentOption
from x402.http.middleware.fastapi import PaymentMiddlewareASGI
from x402.http.types import RouteConfig
from x402.mechanisms.evm.exact import ExactEvmServerScheme
from x402.schemas import Network
from x402.schemas.hooks import SettleResultContext, VerifyResultContext
from x402.server import x402ResourceServer

load_dotenv()

logging.basicConfig(level=logging.INFO)

# Config
EVM_ADDRESS = os.getenv("EVM_ADDRESS")
EVM_NETWORK: Network = "eip155:84532"  # Base Sepolia
FACILITATOR_URL = os.getenv("FACILITATOR_URL", "https://x402.org/facilitator")

if not EVM_ADDRESS:
    raise ValueError("Missing required EVM_ADDRESS environment variable")


class WeatherReport(BaseModel):
    weather: str
    temperature: int


class WeatherResponse(BaseModel):
    report: WeatherReport


app = FastAPI()

facilitator = HTTPFacilitatorClient(FacilitatorConfig(url=FACILITATOR_URL))
server = x402ResourceServer(facilitator)
server.register(EVM_NETWORK, ExactEvmServerScheme())
server.register_extension(bazaar_resource_server_extension)


def _log_extension_responses(phase: str, ctx: VerifyResultContext | SettleResultContext) -> None:
    """Process facilitator extension sidechannel metadata."""
    extension_responses = ctx.extension_responses
    if not extension_responses:
        print(f"\n=== {phase}: no facilitator extension responses ===")
        return

    print(f"\n=== {phase}: facilitator extension responses (sidechannel) ===")
    print(json.dumps(extension_responses, indent=2))

    bazaar = extension_responses.get("bazaar")
    if not isinstance(bazaar, dict):
        return

    status = bazaar.get("status")
    if status == "success":
        print("   ✅ Bazaar cataloging confirmed by facilitator")
    elif status == "processing":
        print("   ⏳ Bazaar cataloging in progress")
    elif status == "rejected":
        reason = bazaar.get("rejectedReason", "unknown")
        print(f"   ⚠️  Bazaar rejected: {reason}")


server.on_after_verify(lambda ctx: _log_extension_responses("After verify", ctx))
server.on_after_settle(lambda ctx: _log_extension_responses("After settle", ctx))

routes = {
    "GET /weather": RouteConfig(
        accepts=[
            PaymentOption(
                scheme="exact",
                pay_to=EVM_ADDRESS,
                price="$0.001",
                network=EVM_NETWORK,
            ),
        ],
        description="Weather data",
        mime_type="application/json",
        service_name="Weather API",
        tags=["weather", "api"],
        extensions={
            **declare_discovery_extension(
                input={"city": "San Francisco"},
                input_schema={
                    "properties": {"city": {"type": "string"}},
                    "required": ["city"],
                },
                output=OutputConfig(
                    example={"weather": "sunny", "temperature": 70},
                    schema={
                        "properties": {
                            "weather": {"type": "string"},
                            "temperature": {"type": "number"},
                        },
                        "required": ["weather", "temperature"],
                    },
                ),
            )
        },
    ),
}
app.add_middleware(PaymentMiddlewareASGI, routes=routes, server=server)


@app.get("/weather")
async def get_weather(city: str = "San Francisco") -> WeatherResponse:
    return WeatherResponse(report=WeatherReport(weather="sunny", temperature=70))


if __name__ == "__main__":
    import uvicorn

    uvicorn.run(app, host="0.0.0.0", port=4021)

File: examples/typescript/servers/advanced/bazaar.ts
import { config } from "dotenv";
import express from "express";
import { paymentMiddleware, x402ResourceServer } from "@x402/express";
import { ExactEvmScheme } from "@x402/evm/exact/server";
import { HTTPFacilitatorClient } from "@x402/core/server";
import { declareDiscoveryExtension } from "@x402/extensions/bazaar";
config();

const evmAddress = process.env.EVM_ADDRESS as `0x${string}`;
if (!evmAddress) {
  console.error("Missing required environment variables");
  process.exit(1);
}

const facilitatorUrl = process.env.FACILITATOR_URL;
if (!facilitatorUrl) {
  console.error("❌ FACILITATOR_URL environment variable is required");
  process.exit(1);
}
const facilitatorClient = new HTTPFacilitatorClient({ url: facilitatorUrl });

const app = express();

app.use(
  paymentMiddleware(
    {
      "GET /weather": {
        accepts: {
          scheme: "exact",
          price: "$0.001",
          network: "eip155:84532",
          payTo: evmAddress,
        },
        description: "Weather data",
        mimeType: "application/json",
        extensions: {
          ...declareDiscoveryExtension({
            input: { city: "San Francisco" },
            inputSchema: {
              properties: {
                city: { type: "string" },
              },
              required: ["city"],
            },
            output: {
              example: {
                city: "San Francisco",
                weather: "foggy",
                temperature: 60,
              },
            },
          }),
        },
      },
    },
    new x402ResourceServer(facilitatorClient).register("eip155:84532", new ExactEvmScheme()),
  ),
);

app.get("/weather", (req, res) => {
  const city = (req.query.city as string) || "San Francisco";

  const weatherData: Record<string, { weather: string; temperature: number }> = {
    "San Francisco": { weather: "foggy", temperature: 60 },
    "New York": { weather: "cloudy", temperature: 55 },
  };

  const data = weatherData[city] || { weather: "sunny", temperature: 70 };

  res.send({
    city,
    weather: data.weather,
    temperature: data.temperature,
  });
});

app.listen(4021, () => {
  console.log(`Server listening at http://localhost:${4021}`);
});
